mlir/lib/Analysis/BufferDependencyAnalysis.cpp

```diff
@@ -0,0 +1,153 @@
+//===- BufferDependencyAnalysis.h - a set of useful generic analyses
+//---------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/BufferDependencyAnalysis.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+#include <optional>
+
+using namespace mlir;
+
+struct UserWrapper {
+  Operation *parentAlias;
+  Operation *op;
+};
+
+void findAllUsers(Operation *parent, llvm::SmallVector<UserWrapper> &users) {
+  for (auto *user : parent->getUsers()) {
+    if (auto viewOp = dyn_cast<ViewLikeOpInterface>(user)) {
+      findAllUsers(viewOp, users);
+    } else {
+      users.push_back({parent, user});
+    }
+  }
+}
+
+enum EffectType { read, write, unknown };
+EffectType getMemoryEffect(UserWrapper &user) {
+  if (auto memoryEffect = dyn_cast<MemoryEffectOpInterface>(user.op)) {
+    int64_t index = -1;
+    for (auto it : llvm::enumerate(user.op->getOperands())) {
+      auto operand = it.value();
+      if (operand.getDefiningOp() == user.parentAlias) {
+        index = static_cast<int64_t>(it.index());
+        break;
+      }
+    }
+    if (index < 0)
+      return EffectType::unknown;
+    Value value = user.op->getOperand(index);
+    if (memoryEffect.getEffectOnValue<MemoryEffects::Write>(value)) {
+      return EffectType::write;
+    }
+    if (memoryEffect.getEffectOnValue<MemoryEffects::Read>(value)) {
+      return EffectType::read;
+    }
+  }
+  return EffectType::unknown;
+}
+
+template <typename Impl>
+void assign(EffectType effect, UserWrapper &user, Impl &impl) {
+  if (effect == Impl::type) {
+    impl.push_back(user.op);
+  }
+}
+
+template <typename Impl, typename... Impls>
+void assign(EffectType effect, UserWrapper &user, Impl &impl, Impls &...impls) {
+  assign(effect, user, impl);
+  assign(effect, user, impls...);
+}
+
+template <typename Impl, typename... Impls>
+void find(memref::AllocOp allocOp, Impl &impl, Impls &...impls) {
+  BufferDependencyAnalysis::Pair result;
+
+  llvm::SmallVector<UserWrapper> users;
+  findAllUsers(allocOp, users);
+  for (auto &user : users) {
+    EffectType memoryEffect = getMemoryEffect(user);
+    assign(memoryEffect, user, impl, impls...);
+  }
+}
+
+struct Readers : public llvm::SmallVector<Operation *> {
```

Comment:
I ... don't like this.

Just have the function take a `SmallVectorImpl<Operation*>&` and a `EffectType` as regular parameters?

Reply:
I don't know. I was quite happy with this implementation. You extend basically `llvm::SmallVector<Operation *>` with just a static field. If I remember correctly the data object layout is not going to change. Plus, it is an internal structure regarding `BufferDependencyAnalysis.cpp` file. So, nobody else is suppose to use it.

You actually need to associate `SmallVectorImpl<Operation*>&` with some `EffectType` value to me the following work

https://github.com/ROCm/rocMLIR/blob/8fc61db4d3638e8b647519598273cdd2488554c2/mlir/lib/Analysis/BufferDependencyAnalysis.cpp#L73-L83

https://github.com/ROCm/rocMLIR/blob/8fc61db4d3638e8b647519598273cdd2488554c2/mlir/lib/Analysis/BufferDependencyAnalysis.cpp#L106-L117

The alternative would be:

```cpp
struct Reader {
  SmallVectorImpl<Operation*>& container;
  static inline constexpr EffectType type = EffectType::read;
}

struct Writer {
  SmallVectorImpl<Operation*>& container;
  static inline constexpr EffectType type = EffectType::write;
}
```
I don't like that `SmallVectorImpl<Operation*>& container;` gets repeated in both auxiliary structures

Reply:
Why do we need this generic `assign()` at all? There's like, two or three cases here? We can just hand-route them to `readers` or `writers` as needed?

Reply:
(To be clear, this isn't a "I won't merge the way you did it", it's a "hold on, the surrounding codebase doesn't pull this trick anywhere I've seen it, are we *sure* we want to pull in this new idiom?")

Reply:
Why do we even need EffectType here ? 

As in would we have a Reader that have some other effect type at all ?

Reply:
@manupak @krzysz00, I completely changed the implementation.

Reply:
resolved